torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -63,197 +64,154 @@ static int64_t count_specified_dimensions(PyObject* index) {
+    throw IndexError("too many indices for tensor of dimension %d", (int)(self_sizes.size()));
   }
 
   Variable result = self;
-  for (int64_t i = 0; i < size; i++) {
+  for (int64_t i = 0; i < PyTuple_GET_SIZE(index); i++) {
     PyObject* obj = PyTuple_GET_ITEM(index, i); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
-    if (THPUtils_checkLong(obj)) {
-      result = applySelect(result, dim, obj, i);
-    } else if (PySlice_Check(obj)) {
-      result = applySlice(result, dim, obj);
-      dim++;
-    } else if (obj == Py_Ellipsis) {
-      dim += self.dim() - specified_dims;
-    } else if (obj == Py_None) {
-      result = result.unsqueeze(dim);
-      dim++;
-    } else if (PyBool_Check(obj)) {
-      result = result.unsqueeze(dim);
-      handle_var(boolToIndexingTensor(result, obj == Py_True)); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
-    } else if (THPVariable_Check(obj)) {
-      auto& var = THPVariable_Unpack(obj);
-      auto scalar_type = var.scalar_type();
-      if (var.dim() == 0 && at::isIntegralType(scalar_type, /*includeBool=*/true)) {
-        if (scalar_type != at::kByte && scalar_type != at::kBool) {
-          result = applySelect(result, dim, obj, i);
-        } else {
-          result = result.unsqueeze(dim);
-          if(scalar_type == at::kBool) {
-            handle_var(boolToIndexingTensor(result, var.item<bool>() != 0));
-          } else {
-            handle_var(boolToIndexingTensor(result, var.item<uint8_t>() != 0));
+
+    auto tensor_index = ([&]() {
+      if (THPUtils_checkLong(obj)) {
+        if (is_tracing && THPVariable_Check(obj)) {
+          recordSelectTrace(THPVariable_Unpack(obj));
+        }
+        return at::indexing::TensorIndex(THPUtils_unpackLong(obj));
+      } else if (PySlice_Check(obj)) {
+        Py_ssize_t start, stop, step;
+        if (!THPUtils_unpackSlice(obj, &start, &stop, &step)) {
+          throw python_error();
+        }
+        if (is_tracing) {
+          Tensor start_tensor, stop_tensor, step_tensor;
+          extractTensorsFromSlice(obj, &start_tensor, &stop_tensor, &step_tensor);
+          recordSliceTrace(start_tensor, stop_tensor, step_tensor);
+        }
+        return at::indexing::TensorIndex({start, stop, step});
+      } else if (obj == Py_Ellipsis) {
+        return at::indexing::TensorIndex(at::indexing::Ellipsis);
+      } else if (obj == Py_None) {
+        return at::indexing::TensorIndex(at::indexing::None);
+      } else if (PyBool_Check(obj)) {
+        return at::indexing::TensorIndex(obj == Py_True);
+      } else if (THPVariable_Check(obj)) {
+        if (is_tracing) {
+          Tensor tensor = THPVariable_Unpack(obj);
+          auto scalar_type = tensor.scalar_type();
+          if (tensor.dim() == 0 && at::isIntegralType(scalar_type, /*includeBool=*/false) && scalar_type != at::kByte) {
+            recordSelectTrace(tensor);
           }
         }
+        return at::indexing::TensorIndex(THPVariable_Unpack(obj));
```

Comment:
You are pointlessly unpacking the tensor twice. Just use `tensor` in the second case. Ideally, move it into `TensorIndex`.